asllib/types.ml

```diff
@@ -631,36 +579,34 @@ let rec bitfields_included env bfs1 bfs2 =
         | BitField_Type (name1, slices1, ty1) ->
             String.equal name1 name2
             && slices_equal env slices1 slices2
-            && structural_subtype_satisfies env ty1 ty2)
+            && subtype_satisfies env ty1 ty2)
   and incl_bfs bfs1 bfs2 = List.for_all (mem_bfs bfs2) bfs1 in
   incl_bfs bfs1 bfs2
 
-(* Begin TypingRule.StructuralSubtypeSatisfaction *)
-and structural_subtype_satisfies env t s =
+(* Begin TypingRule.SubtypeSatisfaction *)
+and subtype_satisfies env t s =
   (* A type T subtype-satisfies type S if and only if all of the following
      conditions hold: *)
   (match ((make_anonymous env s).desc, (make_anonymous env t).desc) with
   (* If S has the structure of an integer type then T must have the structure
      of an integer type. *)
-  | T_Int _, T_Int _ -> true
-  | T_Int _, _ -> false
-  (* If S has the structure of a real type then T must have the
-     structure of a real type. *)
-  | T_Real, T_Real -> true
-  | T_Real, _ -> false
-  (* If S has the structure of a string type then T must have the
-     structure of a string type. *)
-  | T_String, T_String -> true
-  | T_String, _ -> false
-  (* If S has the structure of a boolean type then T must have the
-     structure of a boolean type. *)
-  | T_Bool, T_Bool -> true
-  | T_Bool, _ -> false
+  | T_Int _, T_Int _ ->
+      let d_s = Domain.of_type env s and d_t = Domain.of_type env t in
+      let () =
+        if false then
+          Format.eprintf "domain_subtype_satisfies: %a included in %a?@."
+            Domain.pp d_t Domain.pp d_s
+      in
+      Domain.is_subset env d_t d_s
+  (* If S has the structure of a real/string/bool then T must have the
+     same structure. *)
+  | ( ((T_Real | T_String | T_Bool) as s_anon),
+      ((T_Real | T_String | T_Bool) as t_anon) ) ->
+      s_anon = t_anon
   (* If S has the structure of an enumeration type then T must have
      the structure of an enumeration type with exactly the same
      enumeration literals. *)
   | T_Enum li_s, T_Enum li_t -> list_equal String.equal li_s li_t
-  | T_Enum _, _ -> false
   (*
       • If S has the structure of a bitvector type with determined width then
```

Comment:
Please remove the two conditions in the comments about bitvectors of undetermined widths, which are no longer relevant as bitvectors of undetermined width have been eliminated from ASLv1.